### PR TITLE
Add test.support.busy_retry()

### PR DESCRIPTION
**Pull request by vstinner**
Add busy_wait() and increasing_sleep() functions to test.support.

### CODE REVIEW

**Comment by vstinner** [comment]
cc @erlend-aasland @corona10 @serhiy-storchaka @methane

**Comment by vstinner** [comment]
> "Configure ccache action" failed with: "Error: reserveCache failed: Cache service responded with 503"

https://github.com/actions/cache/issues/820

**Comment by vstinner** [comment]
When writing this PR, I saw code that I wanted to change, but I tried to minimize changes of this PR :-D I will write a follow-up PR to enhance tests.

**Comment by serhiy-storchaka** [comment]
Nice. But some new code is not equivalent to the old code (in particularly related to signal handling). In some cases the old code used very short sleep or no sleep, or other operation instead of sleep. I am not sure that replacements preserve the intention of tests.

**Comment by vstinner** [comment]
> But some new code is not equivalent to the old code (in particularly related to signal handling).

Oh sorry, I reverted these changes. Are remaining changes correct?

File: Lib/test/test_signal.py [hunk 1]
```diff
@@ -812,13 +812,14 @@ def test_itimer_virtual(self):
         signal.signal(signal.SIGVTALRM, self.sig_vtalrm)
         signal.setitimer(self.itimer, 0.3, 0.2)
 
-        start_time = time.monotonic()
-        while time.monotonic() - start_time < 60.0:
+        for _ in support.increasing_sleep(60.0, error=False):
```
Comment (serhiy-storchaka) [inline Lib/test/test_signal.py]: There were no `time.sleep()` in this loop. Can you disable calling `time.sleep()` here?

I suppose there was a reason of using time consuming computation instead of `time.sleep()`.

Comment (vstinner) [reply Lib/test/test_signal.py]: Oops, well spotted. I added busy_wait() for that: I fixed the 2 tests.

**Comment by vstinner** [comment]
I dislike ``increasing_sleep()`` name. What do you think of ``busy_wait_and_sleep()``?

**Comment by vstinner** [comment]
The "busy wait" term seems commonly used: https://en.wikipedia.org/wiki/Busy_waiting

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
> I dislike `increasing_sleep()` name. What do you think of `busy_wait_and_sleep()`?

You can merge both functions if skip `time.sleep()` when `init_delay` is 0.

I though also about alternative interface. Instead of
```
for _ in support.increasing_sleep(support.SHORT_TIMEOUT):
    if check():
        break
```
you could write write
```
support.wait_for(check, timeout=support.SHORT_TIMEOUT)
```
(looks like `Condition.wait_for()`)
or
```
w = support.expired(support.SHORT_TIMEOUT)
while not check() and not expired:
    pass
```
(it is closer to the original code).

But I do not want to start bikeshedding. It is all wild fantasy. These functions are for internal use. The current code of the PR is good enough.

**Comment by methane** [comment]
bikeshedding: busy_retry and sleeping_retry.

**Comment by vstinner** [comment]
> support.wait_for(check, timeout=support.SHORT_TIMEOUT)

First, I wanted to do something like that, but I dislike defining nested functions or lambda functions. If you read the PR, you can see that existing tests look the same with my change. There is no indirection (function call), the code remains "flat".

**Comment by vstinner** [comment]
increasing_sleep() is a bad name. The delay is not "increasing", it's exponential. What the function does is more an: https://en.wikipedia.org/wiki/Exponential_backoff

IMO tenacity API https://tenacity.readthedocs.io/en/latest/api.html#wait-functions should be a good inspiration for function names. I renamed increasing_sleep() to wait_exponential().

**Comment by serhiy-storchaka** [comment]
> bikeshedding: busy_retry and sleeping_retry.

Yes, I would expect the word "repeat" in the name too. But I do not want to make it a requirement.

> First, I wanted to do something like that, but I dislike defining nested functions or lambda functions. If you read the PR, you can see that existing tests look the same with my change. There is no indirection (function call), the code remains "flat".

I understood you. While in some cases it is easy to write a lambda, in other cases, when you use several statement in the loop or even try/except block, the local function is unavoidable, and it makes the code less straight.

> increasing_sleep() is a bad name. The delay is not "increasing", it's exponential.

I think that the name should describe the purpose of the function, not the implementation detail. In some cases linear grow of delays (and quadratic of total time) may be more preferable than exponential grow.

What do you think about merging two functions into a single function?

**Comment by vstinner** [comment]
IMO the two functions are too different and deserve to have different names. In terms of API, currently, I propose:

* busy_wait(timeout)
* wait_exponential(timeout) -- was called increasing_sleep(timeout) in the first version

@methane proposed:

* busy_retry(timeout)
* sleeping_retry(timeout)

tenacity API seems to be:

* stop_after_delay(timeout)
* wait_exponential(max=timeout)

I like to put "sleep" in the name, so it's easy to understand that the function calls time.sleep(). I have no preference between "retry and "wait". Maybe retry better highlights that it's a loop and it calls the loop body multiple times until it "succeed" (somehow).

In short, I like names proposed by @methane. @serhiy-storchaka: do you have a preference?

By the way, the whole API is:

```
wait_exponential(timeout, err_msg=None, /, *, init_delay=0.001, max_delay=1.0, error=True)
```

Are you fine with init_delay and max_delay parameter names? If not, do you have a suggestion?

Well, I'm not sure that the API matters so much, it's a test module :-D It must not be used outside Python ;-)

**Comment by serhiy-storchaka** [comment]
I like @methane proposal.
